pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -82,6 +83,58 @@ const (
 	imageDigestExporterContainerName = "step-image-digest-exporter"
 )
 
+type TaskRunWithTask struct {
+	taskRunSteps []v1alpha1.StepState
+	mapForSort map[string]int
+}
+
+func (trt *TaskRunWithTask) Init(taskRunSteps []v1alpha1.StepState, taskSpecSteps []corev1.Container) {
+	trt.taskRunSteps = taskRunSteps
+	trt.mapForSort = trt.constructTaskStepsSorter(taskSpecSteps)
+}
+
+// This function constructTaskStepsSorter constructs a map matching the names of
+// the steps to their indices for a task.
+func (trt *TaskRunWithTask) constructTaskStepsSorter(taskSpecSteps []corev1.Container) map[string]int {
+	sorter := make(map[string]int)
+	for index, step := range taskSpecSteps {
+		sorter[step.Name] = index
+	}
+	return sorter
+}
+
+// This function changeIndex sorts the steps of the task run, based on the
+// order of the steps in the task. Instead of changing the element with the one next to it,
+// we directly swap it with the desired index.
+func (trt *TaskRunWithTask) changeIndex(index int) {
+	// Check if the current index is equal to the desired index. If they are equal, do not swap; if they
+	// are not equal, swap index j with the desired index.
+	desiredIndex, exist := trt.mapForSort[trt.taskRunSteps[index].Name]
+	if exist && index != desiredIndex {
+		trt.taskRunSteps[desiredIndex], trt.taskRunSteps[index] = trt.taskRunSteps[index], trt.taskRunSteps[desiredIndex]
+	}
+}
+
+func (trt *TaskRunWithTask) Len() int {
+	return len(trt.taskRunSteps)
+}
+
+func (trt *TaskRunWithTask) Swap(i, j int) {
+	trt.changeIndex(j)
+	// The index j is unable to reach the last index.
+	// When i reaches the end of the array, we need to check whether the last one needs a swap.
+	if (i == trt.Len() - 1 ) {
+		trt.changeIndex(i)
+	}
+}
+
+func (trt *TaskRunWithTask) Less(i, j int) bool {
+	// Since the logic is complicated, we move it into the Swap function to decide whether
+	// and how to change the index. We set it to true here in order to iterate all the
+	// elements of the array in the Swap function.
+	return true
+}
+
```

Comment:
Would you mind we have a new package with a more general name like utils, so that we can put this sorter and maybe further algorithm(TBD) inside it?

Reply:
I'd like to name a new package utils.

Reply:
@houshengbo I would definitely advise against `utils` :sweat_smile: generic names like `utils` or `helpers` do not convey anything of what is inside. This function is about sorting taskrun, so it could either be called `sort` (inside `taskrun`) or something a tiny bit more general like `list` ?

Reply:
(also even if we would keep `pkg/utils`, it should *not* depend on the api types)

Reply:
An example on `tektoncd/cli` : `tasks.SortTasksBySpecOrder`

Reply:
Forget about my previous comment.
what about stepsorting

Reply:
@houshengbo :hugs: as it's related to `tasks` I feel it would be better in a package name that has `task` in (the path or the name) (`tasks.SortBySpecOrder` instead of my previous propositon)

cc @bobcatfish as I am not the greatest to find correct names :joy:

Reply:
ooo i like both the ideas `stepsorting` is pretty close to what I would have suggested but I think @vdemeester is right, we'd end up with something like `stepsorting.Sort` 🤔 

we have been kind of in need of a `status` related package for a while now (we have lots have status stuff in the reconciler) how about something like `taskrunstatus` or `status`, so we could have `taskrunstatus.SortSteps` or `status.SortSteps`?

Reply:
ohh I do like the `status` package :heart_eyes_cat: `status.SortSteps` sounds **really good** to me !

Reply:
yep, I think we can agree on that.